util/topgen/c.py

```diff
@@ -463,3 +464,24 @@ def _init_clkmgr_clocks(self):
 
         self.clkmgr_gateable_clocks = gateable_clocks
         self.clkmgr_hintable_clocks = hintable_clocks
+
+    def _init_mmio_region(self):
+        """
+        Computes the bounds of the MMIO region.
+
+        The MMIO region excludes ROM, RAM, and flash but includes retention RAM.
+        """
+        RET_RAM_NAME = "sram_ctrl_ret_aon"
```

Comment:
is there any way we can avoid hardcoding these?
since the sram modules have a `memory` key, could we use that instead to disambiguate?

Reply:
Here's what I came up with:
- the most significant nibble of the address determines the address space of a peripheral, i.e. `mask = 0xf0000000`.
- most popular address space is the module configuration space.
- Any peripheral within that address space is considered in the MMIO region.

WDYT?

Reply:
hmmm that would work. But it would be nice if we didn't have to assume and could directly calculate this from the hjson.

Reply:
Yeah but based on the comments in the linked issue I thought the HJSON file didn't have this information

Reply:
wait sorry, i think i might be missing something. What information is missing..? let me look at the issue.

Reply:
sorry i think i was behind on this, that's why i didn't understand what you were saying.
Looking back to your suggestion, what did you mean by `most popular address space`? 
For example if i had a block that had the following..

` base_addrs: {core: "0x40130000", mem: "0x80132000"}`, would the higher address end up being considered the configuration space?

Reply:
what i meant by checking against the attributes, is mostly that we can determine via the `memory` field that something is meant to be a memory and not part of the configuration space.

like below .. 
```
      memory: {
        ram: {
          label:    "ram_ret_aon",
          swaccess:   "rw",
          data_intg_passthru: "true",
          exec:       "True",
          byte_write: "True",
          size:     "0x1000"
        }
      }```

Reply:
Correct, but ret_ram breaks that pattern. That's why I had to hardcode its name.

Reply:
ah yes, but i think you can cross check against the `memory` key.  That will tell you which interface actually belongs in the configuration space, and which interface is a "memory" and can be ignored.

Reply:
I see, thanks for the example and explanation! But we would like to include the ret_ram memory in the mmio address space because it's right there even though it's marked as memory. Otherwise we would have to use more ePMP regions to handle the MMIO and ret_ram address spaces separately, right? 

We could just ignore this and expect things to work out but if ret_ram memory base addr somehow ends up being the largest address, we would end up excluding it from the MMIO region. (Tests would fail if ROM cannot access ret_ram so we would be able to catch this.)

I guess our definition of MMIO is not super consistent since it also includes the ret_ram memory and hence the confusion.

Reply:
Ah, alternatively I can just revert back to the 1st version, remove the hardcoded ret_ram stuff and add an assertion in c code to check this assumption (ret_ram memory must be in the MMIO address space - basically I'll move the asserts in the python code to C)

Reply:
ah! i see. Then okay, perhaps hardcoding is unavoidable then :) 
Maybe let's go with that then.  Then we'll just have to tweak slightly for each top depending on how the memory is assigned.

Reply:
Thanks @tjaychen! Created #14345.

Reply:
Updated as we discussed:
* Moved the assertions to the c code
* Simplified the topgen change to use `TopGenC::memories()` except for `rv_dm` (see #14345).